extensions/omni_sql/omni_sql.c

```diff
@@ -606,147 +716,44 @@ Datum execute_parameterized(PG_FUNCTION_ARGS) {
-            call_ctx->types[i] = TEXTOID;
-            call_ctx->values[i] =
-                PointerGetDatum(cstring_to_text_with_len(val.val.string.val, val.val.string.len));
-            break;
-          case jbvNull:
-            break;
-          default:
-            ereport(ERROR, errmsg("unsupported parameter type at index %i", i));
-          }
-          i++;
-        }
-      }
-
-      // Check types
+      extract_information_json_parameters(call_ctx, funcctx, PG_GETARG_JSONB_P(1));
       if (!PG_ARGISNULL(2)) {
-        ArrayType *types = PG_GETARG_ARRAYTYPE_P(2);
-        ArrayIterator iter = array_create_iterator(types, 0, NULL);
-        int i = 0;
-        Datum type_val;
-        bool isnull;
-        while (array_iterate(iter, &type_val, &isnull)) {
-          if (isnull) {
-            continue;
-          }
-          Oid id = DatumGetObjectId(type_val);
-          // Handle numeric type specialization
-          if (call_ctx->types[i] == NUMERICOID) {
-            switch (id) {
-            case INT2OID:
-              call_ctx->values[i] =
-                  Int16GetDatum(numeric_int4_opt_error(DatumGetNumeric(call_ctx->values[i]), NULL));
-              break;
-            case INT4OID:
-              call_ctx->values[i] =
-                  Int32GetDatum(numeric_int4_opt_error(DatumGetNumeric(call_ctx->values[i]), NULL));
-              break;
-            case INT8OID:
-#if PG_MAJORVERSION_NUM >= 17
-              call_ctx->values[i] =
-                  Int64GetDatum(numeric_int8_opt_error(DatumGetNumeric(call_ctx->values[i]), NULL));
-#else
-              call_ctx->values[i] = DirectFunctionCall1(numeric_int8, call_ctx->values[i]);
-#endif
-              break;
-            case FLOAT4OID:
-              call_ctx->values[i] = DirectFunctionCall1(numeric_float4, call_ctx->values[i]);
-              break;
-            case FLOAT8OID:
-              call_ctx->values[i] = DirectFunctionCall1(numeric_float8, call_ctx->values[i]);
-              break;
-            default: {
-              HeapTuple typeTuple = SearchSysCache1(TYPEOID, ObjectIdGetDatum(id));
-              if (!HeapTupleIsValid(typeTuple)) {
-                ereport(ERROR, (errmsg("cache lookup failed for type %u", id)));
-              }
-
-              char *typename = pstrdup(NameStr(((Form_pg_type)GETSTRUCT(typeTuple))->typname));
-
-              ReleaseSysCache(typeTuple);
-
-              ereport(ERROR, errmsg("can't convert numeric to type %s", typename));
-              break;
-            }
-            }
-          }
-          call_ctx->types[i] = id;
-          i++;
-        }
-        array_free_iterator(iter);
+        extract_information_json_types(call_ctx, PG_GETARG_ARRAYTYPE_P(2));
       }
     }
-
+    SPI_connect();
+    // Prepare the cursor
+    SPIPlanPtr plan = SPI_prepare(call_ctx->stmt, call_ctx->nargs, call_ctx->types);
+    // is this a plan that could return rows?
+    call_ctx->returning = SPI_is_cursor_plan(plan);
```

Comment:
Interesting idea. The difference in approach here is that `SPI_is_cursor_plan` looks at the plan, whereas `omni_sql_is_returning_statement` looks at the syntax tree, and it doesn't have to be valid.

So, outside of this content, `omni_sql_is_returning_statement` is useful for arbitrary queries. Here, `SPI_is_cursor_plan` might be an option to consider because it may be a tiny bit faster (though  `omni_sql_is_returning_statement` doesn't do much either). So it's kind of a toss for me. Thoughts?

Reply:
The reason I did it was because it seemed clearer (fewer lines and less indirection) and probably faster. It would also make the function `omni_sql_is_returning_statement` stale, allowing it to be removed.

Reply:
`omni_sql_is_returning_statement` might be useful as a standalone function for SQL API – but just how often it can be useful is another question. 

At the call site, it's definitely simpler (and a tiny bit faster) to use `SPI_is_cursor_plan` because we won't need to prepare the query.